aws-lc-rs/src/agreement.rs

```diff
@@ -218,6 +230,21 @@ impl EphemeralPrivateKey {
                 rng.fill(&mut priv_key)?;
                 Self::from_p384_private_key(&priv_key)
             }
+            AlgorithmID::ECDH_P521 => {
+                let mut priv_key = [0u8; ECDH_P521_PRIVATE_KEY_LEN];
+                // If the rng is only used for testing purposes, we set the hard-coded value. This is the case
+                // for our KATs. Otherwise, we delegate the random key generation to AWS-LC.
+                //
+                // P-521 is not a Montgomery curve, so we can't directly generate a random key and assign it to
```

Comment:
This is not the issue. The problem is that you are generating a private key that's too large. The order of the P-521 group `n` is a number that's 521-bits long and the private key has to be smaller than `n`. By generating 66 random bytes you are generating a random 528-bit number. With very high probability this number will be larger than `n`, in which case `EC_KEY_set_private_key` will fail. Exactly the same applies to the other curves, but for P-521 it happens with much larger probability.

I see two proper ways to handle this:
1. Relegate the key generation to AWS-LC for all curves.
2. Generate a valid random private key for all curves (`0 < priv_key < n`).